witchcraft-server/src/service/tls.rs

```diff
@@ -15,158 +15,176 @@ use crate::service::hyper::NewConnection;
+                store.add_parsable_certificates(&certs);
+                builder.with_client_cert_verifier(Arc::new(
+                    AllowAnyAnonymousOrAuthenticatedClient::new(store),
+                ))
+            }
+            None => builder.with_no_client_auth(),
+        };
 
-            builder.set_verify(SslVerifyMode::PEER);
-        }
+        let cert_chain = load_certificates(config.keystore().cert_path())?
+            .into_iter()
+            .map(Certificate)
+            .collect();
 
-        let cipher_list = convert_suites(CIPHER_SUITES);
-        builder
-            .set_cipher_list(&cipher_list)
-            .map_err(Error::internal_safe)?;
+        let key_der = load_private_key(config.keystore().key_path())?;
 
-        let ciphersuites = convert_suites(TLS13_CIPHER_SUITES);
-        builder
-            .set_ciphersuites(&ciphersuites)
+        let mut server_config = builder
+            .with_single_cert(cert_chain, key_der)
             .map_err(Error::internal_safe)?;
 
+        server_config.ignore_client_order = true;
         if config.server().http2() {
-            builder.set_alpn_select_callback(|_, client| {
-                ssl::select_next_proto(b"\x02h2\x08http/1.1", client).ok_or(AlpnError::ALERT_FATAL)
-            });
+            server_config.alpn_protocols = vec![b"h2".to_vec(), b"http/1.1".to_vec()];
         }
 
-        builder.set_options(SslOptions::CIPHER_SERVER_PREFERENCE);
-        builder
-            .set_min_proto_version(Some(SslVersion::TLS1_2))
-            .map_err(Error::internal_safe)?;
-        builder
-            .set_max_proto_version(Some(SslVersion::TLS1_3))
-            .map_err(Error::internal_safe)?;
-
-        let mut session_id_ctx = [0; 32];
-        rand::rand_bytes(&mut session_id_ctx).map_err(Error::internal_safe)?;
-        builder
-            .set_session_id_context(&session_id_ctx)
-            .map_err(Error::internal_safe)?;
-
         Ok(TlsLayer {
-            context: builder.build(),
+            acceptor: TlsAcceptor::from(Arc::new(server_config)),
         })
     }
 }
 
+fn load_certificates(path: &Path) -> Result<Vec<Vec<u8>>, Error> {
+    let file = File::open(path).map_err(Error::internal_safe)?;
+    let mut reader = BufReader::new(file);
+    rustls_pemfile::certs(&mut reader).map_err(Error::internal_safe)
+}
+
+fn load_private_key(path: &Path) -> Result<PrivateKey, Error> {
+    let file = File::open(path).map_err(Error::internal_safe)?;
+    let mut reader = BufReader::new(file);
+
+    let mut items = rustls_pemfile::read_all(&mut reader).map_err(Error::internal_safe)?;
+
+    if items.len() != 1 {
+        return Err(Error::internal_safe(
+            "expected exactly one private key in key file",
+        ));
+    }
+
+    match items.pop().unwrap() {
+        Item::RSAKey(buf) | Item::PKCS8Key(buf) | Item::ECKey(buf) => Ok(PrivateKey(buf)),
+        _ => Err(Error::internal_safe(
```

Comment:
nit: Not sure if there's an easy way to get the variant, but may be worth including the type as an unsafe arg?

Reply:
Unfortunately not in a stringified fashion.